cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -224,13 +228,27 @@ protected void shutDown() throws Exception {
     List<ApplicationDetail> result = new ArrayList<>();
     for (Map.Entry<ApplicationId, ApplicationSpecification> entry : appSpecs.entrySet()) {
       ApplicationDetail applicationDetail = ApplicationDetail.fromSpec(entry.getValue(), owners.get(entry.getKey()));
+      try {
+        ensureApplicationEnabled(entry.getValue());
+      } catch (CapabilityNotAvailableException ex) {
+        LOG.debug("Application {} is ignored due to exception.", applicationDetail.getName(), ex);
+        continue;
+      }
       if (predicate.test(applicationDetail)) {
         result.add(filterApplicationDetail(entry.getKey(), applicationDetail));
       }
     }
     return result;
   }
 
+  private void ensureApplicationEnabled(ApplicationSpecification appSpec) throws IOException,
+    CapabilityNotAvailableException {
+    for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
```

Comment:
Repeated logic as in the `LocalArtifactLoaderStage` class? Consider moving it into `CapabilityReader.checkAllEnabled(ApplicationSpecification spec)`.

Reply:
Moved